docs/madr/decisions/019-offline-signing-tokens.md

```diff
@@ -0,0 +1,207 @@
+*validator.mapping* - list of fields in claims to map. Users can provide their own tokens with fields that are different from what we require.
+
+#### Use cases
+
+**Use tokens issued by the CP**
+```yaml
+apiServer:
+  authn:
+    type: tokens
+    tokens:
+      enableIssuer: true
+      validator:
+        useSecrets: true
+```
+
+**Use tokens issued by the CP and an external issuer. Tokens still can be issued by the CP**
+```yaml
+apiServer:
+  authn:
+    type: tokens
+    tokens:
+      enableIssuer: true
+      validator:
+        useSecrets: true
+        publicKeys:
+        - kid: 123
+          key: ...
+```
+
+**Use tokens issued by the CP and external issuer. Tokens cannot be issued by the CP**
+```yaml
+apiServer:
+  authn:
+    type: tokens
+    tokens:
+      enableIssuer: false
+      validator:
+        useSecrets: true
+        publicKeys:
+          - kid: 123
+            key: ...
+```
+
+**Use tokens issued by external issuer**
+```yaml
+apiServer:
+  authn:
+    type: tokens
+    tokens:
+      enableIssuer: false
+      validator:
+        useSecrets: false
+        publicKeys:
+          - kid: 123
+            key: ...
+```
+
+### Dataplane Token configuration
+
+```yaml
+dpServer:
+  authn:
+    dpp:
+      type: tokens # or serviceAccountToken or none
+      tokens:
+        enableIssuer: true
+        validator:
+          useSecrets: true
+          publicKeys:
+            - kid: 123
+              key: ...
+              mesh: default # <- DPP token is mesh scoped
+```
+
+We won't introduce `mapping` yet as it's very unlikely at this moment that a user would like to map it from existing token.
+
+Additionally, we deprecate this config
+```yaml
+dpServer:
+  auth:
+    type: "" # ENV: KUMA_DP_SERVER_AUTH_TYPE
+```
+Because it's common for both data plane proxies and zone proxies. Since both uses different way of auth, it should be separate.
+
+### Zone token configuration
+
+```yaml
+dpServer:
+  authn:
+    zoneProxies:
+      type: tokens # or serviceAccountToken or none
+      tokens:
+        enableIssuer: true
+        validator:
+          useSecrets: true
+          publicKeys:
+            - kid: 123
+              key: ...
+```
+
+
+### Observability
+
+We need to introduce a metric to see which `kid` was used to validate the token.
+This will help users to know when they can migrate signing keys.
+
+### kumactl
+
+`kumactl` right now offers token generation, but it uses CP connection. To support offline signing, we need to improve `kumactl`.
+
+#### Generating signing key
+
+Currently, we have a command to generate a signing key that does not require CP connection.
+```
+$ kumactl generate signing-key
+LS0tLS1CRUdJTiBSU0EgUFJ........
+```
+However, it produces a private key
+```
+$ kumactl generate signing-key | base64 -d
+-----BEGIN RSA PRIVATE KEY-----
+MIIEogIBAAKCAQEAy0KtfI7O0TJ00............
+-----END RSA PRIVATE KEY-----
+```
+
+We need to have a way of extracting public key from it.
```

Comment:
I'm guessing this is just going to be a wrapper around

```
openssl rsa -in private.pem -out public.pem -outform PEM -pubout
```

right?

Reply:
sort of, but in Go. We already have util functions to do this.